pkg/api/bzzfile.go

```diff
@@ -0,0 +1,214 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package api
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"mime"
+	"mime/multipart"
+	"net/http"
+
+	"github.com/ethersphere/bee/pkg/collection/entry"
+	"github.com/ethersphere/bee/pkg/file"
+	"github.com/ethersphere/bee/pkg/file/joiner"
+	"github.com/ethersphere/bee/pkg/file/splitter"
+	"github.com/ethersphere/bee/pkg/jsonhttp"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/gorilla/mux"
+)
+
+const (
+	MultiPartFormData = "multipart/form-data"
+)
+
+type FileUploadResponse struct {
+	Address swarm.Address `json:"address"`
+}
+
+func (s *server) bzzFileUploadHandler(w http.ResponseWriter, r *http.Request) {
+	contentType, params, err := mime.ParseMediaType(r.Header.Get("Content-Type"))
+	if contentType != MultiPartFormData {
+		s.Logger.Debugf("bzz-file: no mutlipart: %v", err)
+		s.Logger.Error("bzz-file: no mutlipart ")
+		jsonhttp.BadRequest(w, "not a mutlipart/form-data")
+		return
+	}
+
+	mr := multipart.NewReader(r.Body, params["boundary"])
+	for {
+		part, err := mr.NextPart()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			s.Logger.Debugf("bzz-file: read mutlipart: %v", err)
+			s.Logger.Error("bzz-file: read mutlipart ")
+			jsonhttp.BadRequest(w, "error reading a mutlipart/form-data")
+			return
+		}
+
+		cd := part.Header.Get("Content-Disposition")
+		_, params, err := mime.ParseMediaType(cd)
+		if err != nil {
+			s.Logger.Debugf("bzz-file: parse mediatype: %v", err)
+			s.Logger.Error("bzz-file: parse mediatype")
+			jsonhttp.BadRequest(w, "error parsing parse mediatype")
+			return
+		}
+		fileName := params["name"]
+		if fileName == "" {
+			fileName = part.FormName()
+		}
+
+		// set up splitter to process the data
+		splitr := splitter.NewSimpleSplitter(s.Storer)
+		ctx := context.Background()
+
+		// first store the file and get the reference
+		buf, err := ioutil.ReadAll(part)
+		if err != nil {
+			s.Logger.Debugf("bzz-file: read part: %v", err)
+			s.Logger.Error("bzz-file: read part ")
+			jsonhttp.BadRequest(w, "read a part data")
+			return
+		}
+		fr, err := s.storeData(ctx, splitr, buf)
+		if err != nil {
+			s.Logger.Debugf("bzz-file: file store: %v, file name %s", err, fileName)
+			s.Logger.Error("bzz-file: file store")
+			jsonhttp.InternalServerError(w, "could not store metadata")
+			return
+		}
+
+		// then create and store metadata.
+		m := entry.NewMetadata(fileName)
+		contentType := part.Header.Get("Content-Type")
+		if contentType == "" {
+			contentType = http.DetectContentType(buf)
```

Comment:
This magic is the reason why you do the extra buffer? I am (still) of the opinion that this kind of magic is unnecessary for a lean bee node.